MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift

```diff
@@ -0,0 +1,143 @@
+            return dataSource.get(query)
+        }
+    }
+    
+    public func getAll(_ query: Query) -> Future<[T]> {
+        switch query {
+        case let query as KeyQuery:
+            if let cached = arrays[query.key], isValid(cached.date) {
+                return Future(cached.array)
+            }
+            return dataSource.getAll(query).filter { array in
+                self.arrays[query.key] = (array, Date())
+                }.recover { error in
+                    self.arrays[query.key] = nil
+                    return Future(error)
+            }
+        default:
+            print("TimedCacheDataSource can't cache the result of the \(type(of: dataSource)).getAll call because \(type(of: query)) doesn't conform to KeyQuery.")
+            return dataSource.getAll(query)
+        }
+    }
+    
+    @discardableResult
+    public func put(_ value: T? = nil, in query: Query) -> Future<T> {
+        switch query {
+        case let query as KeyQuery:
+            return dataSource.put(value, in: query).filter { object in
+                    self.objects[query.key] = (object, Date())
+                }.recover { error in
+                    return Future(error)
+                }
+        default:
+            print("TimedCacheDataSource can't cache the result of the \(type(of: dataSource)).put call because \(type(of: query)) doesn't conform to KeyQuery.")
+            return dataSource.put(value, in: query)
+        }
+    }
+    
+    @discardableResult
+    public func putAll(_ array: [T], in query: Query) -> Future<[T]> {
+        switch query {
+        case let query as KeyQuery:
+            return dataSource.putAll(array, in: query).filter { array in
+                self.arrays[query.key] = (array, Date())
+                }.recover { error in
+                    return Future(error)
+            }
+        default:
+            print("TimedCacheDataSource can't cache the result of the \(type(of: dataSource)).putAll call because \(type(of: query)) doesn't conform to KeyQuery.")
+            return dataSource.putAll(array, in: query)
+        }
+    }
+    
+    @discardableResult
+    public func delete(_ query: Query) -> Future<Void> {
+        switch query {
+        case let query as KeyQuery:
+            return dataSource.delete(query).filter { _ in
+                self.objects[query.key] = nil
+            }
+        default:
+            print("TimedCacheDataSource can't cache the result of the \(type(of: dataSource)).delete call because \(type(of: query)) doesn't conform to KeyQuery.")
+            return dataSource.delete(query)
+        }
+    }
+    
+    @discardableResult
+    public func deleteAll(_ query: Query) -> Future<Void> {
+        switch query {
+        case let query as KeyQuery:
+            return dataSource.delete(query).filter { _ in
+                self.arrays[query.key] = nil
+            }
```

Comment:
What if error? should we delete the current cached object anyway?

Reply:
We should delete cached values.